File: PP/Lista2/fibTail.ml
(* O(n), int -> int *)
let fibTail n =
    let rec aux (n, (x, y)) = if n = 0 then x else aux (n-1, (y, x+y))
        in if n >= 0 then aux (n, (0, 1)) else failwith "Ujemna liczba";;

print_int (fibTail 1);;

File: PP/Lista2/fibTail.scala
object fibTail {
    def main(args: Array[String]) : Unit = {
        def fibTail(x : Int) : BigInt =
            def aux(x : Int, a : Int, b : Int) : Int =
                if x == 0 then a else aux (x-1, b, a+b)
            if x > 0
                then return aux (x, 0, 1)
                else throw new Exception("Ujemna liczba");
        println(fibTail(42))
    }
}